test/e2e/catalogutil/catalog.go

```diff
@@ -0,0 +1,1537 @@
+	if !ok {
+		return errors.New("Unable to get digest from skopeo inspect")
+	}
+	destinationImageWithDigest := fmt.Sprintf("%s/%s@%s", stack.TargetRegistry, packageNameLower, digest)
+	catalogEntry.BundleImageWithDigest = destinationImageWithDigest
+	return nil
+}
+
+/*
+RegistryAdd is a function that uses "opm registry add" to add a catalog entry to a catalog
+
+Arguments:
+
+• toolsbin: The absolute path to the directory that contains the tools needed to generate a catalog
+
+• destinationDir: The "root destination" directory that contains the manifest files (where the bundles.db will end up)
+
+• catalogEntry: a single catalog entry that contains the information needed to add a bundle image to a catalog. Note that
+catalogEntry.NewIndex controls when a database should be created from scratch before adding an entry
+
+• stack: contains arguments needed to generate a catalog for a specific environment
+
+Returns:
+
+error is non nil if anything prevented the bundle image from being added to the database
+*/
+func RegistryAdd(toolsbin string, destinationDir string, catalogEntry *CatalogEntry, stack Stack) error {
+
+	opm, err := stack.GetOPM()
+	if err != nil {
+		return err
+	}
+
+	// Build using opm registry commands
+	databaseFile := filepath.Join(destinationDir, "bundles.db")
+
+	// if we want a new index, blow away any database file that might be present
+	if catalogEntry.NewIndex {
+		// remove any existing database
+		os.Remove(databaseFile)
+	}
+	cmd := exec.Command(
+		filepath.Join(toolsbin, opm), "registry", "add",
+		"--skip-tls",
+		"-b", catalogEntry.BundleImageWithDigest,
+		"-c", stack.ContainerCLI.String(),
+		"--mode", catalogEntry.Addmode.String(),
+	)
+	if stack.OpmDebug {
+		setOPMDebug(cmd)
+	}
+	cmd.Dir = destinationDir
+	log(fmt.Sprintf("Executing command: %v", cmd))
+	cmdOutputBytes, err := cmd.CombinedOutput()
+	log(string(cmdOutputBytes))
+	if err != nil {
+		return err
+	}
+
+	sqlLiteCommand(databaseFile, "--package--", "select * from package")
+	sqlLiteCommand(databaseFile, "--channel--", "select * from channel")
+	sqlLiteCommand(databaseFile, "--channel-entry--", "select * from channel_entry")
+	sqlLiteCommand(databaseFile, "--operatorbundle--", "select name,bundlepath,replaces,skips,skiprange,length(csv),length(bundle) from operatorbundle;")
+	return nil
+}
+
+/*
+CreateIndexFromDatabase is a function that generates a catalog image using a bundles.db. The
+dockerfile is configured for either upstream or downstream use depending on the values provided in stack.
+
+Arguments:
+
+• destinationDir: The "root destination" directory that contains the manifest files (where the bundles.db is located)
+
+• stack: contains arguments needed to generate a catalog for a specific environment
+
+Returns:
+
+error is non nil if anything prevented the catalog image from being created (NOTE: pushing occurs independently)
+*/
+func CreateIndexFromDatabase(destinationDir string, stack Stack) error {
```

Comment:
This assumes a database is present in the catalog image, which is something we are actively trying to move away from. Testing both existing catalogs and the newer declarative config-based ones would be challenging, but it could be extremely useful. In cases like that, we probably don't want to assume the index is one way or another until we actually unpack and inspect its metadata contents.

Reply:
There's definitely work to do for handling the new catalog style, but I think if we start with this PR we at least cover the existing catalogs. My thinking on this is making the catalog creation function `CreateTemporaryCatalog` use another path for handling the new catalog type by adding a new `TargetCatalogType` with the `Stack`.

Reply:
I agree, @jchunkins